### PR TITLE
Reject empty password in internal user creation

### PR DESCRIPTION
**Pull request by cliu123**
*Issue #, if available:* 

*Description of changes:*
Given that Kibana and ES don't allow login with empty password, it'd make sense to deny empty password in internal user creation.
This PR is to add the validation to deny empty password in internal user creation.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by sujithvm** [comment]
Any reason why we want to allow empty password? Empty password may be considered as security vulnerability and most UI/front end systems block passing empty passwords

File: src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/InternalUsersApiAction.java [hunk 1]
```diff
@@ -116,8 +116,14 @@ protected void handlePut(RestChannel channel, final RestRequest request, final C
             }
         }
 
-        // if password is set, it takes precedence over hash
+        // Reject empty password
         final String plainTextPassword = securityJsonNode.get("password").asString();
+        if (plainTextPassword != null && plainTextPassword.isEmpty()){
+            badRequestResponse(channel, "Please specify a non-empty 'password' when creating a new internal user.");
```
Comment (sujithvm) [inline src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/InternalUsersApiAction.java:122]: Can you reject in CredentialsValidator which is used by the account API for self-change password as well ?

File: src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/InternalUsersApiAction.java [hunk 2]
```diff
@@ -116,8 +116,14 @@ protected void handlePut(RestChannel channel, final RestRequest request, final C
...
+            return;
+        }
+
+        // if password is set, it takes precedence over hash
```
Comment (sujithvm) [inline src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/InternalUsersApiAction.java:126]: Can you remove other checks for `isEmpty()` ?

File: src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/InternalUsersApiAction.java (hunk 1 above)
Comment (cliu123) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/InternalUsersApiAction.java:122]: I tried that. But CredentialsValidator validates regex, and it respondes status code 201: {"status":"error","reason":"Password does not match minimum criteria"}. I believe what we want for empty password is status code 400: {"status":"BAD_REQUEST","message":"Please specify a non-empty 'password' when creating a new internal user."}.

File: src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/InternalUsersApiAction.java (hunk 2 above)
Comment (cliu123) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/InternalUsersApiAction.java:126]: Done.

Comment (sujithvm) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/InternalUsersApiAction.java:126]: Can you simplify the checks? we only need to check for password is not null and replace hash so that it takes precedence

File: src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/InternalUsersApiAction.java (hunk 1 above)
Comment (sujithvm) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/InternalUsersApiAction.java:122]: It returns 201 with that error message? This should not happen. 400 must be returned. Can you verify again

Comment (cliu123) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/InternalUsersApiAction.java]: Moved the validation to CreadentialValidator.

File: src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/CredentialsValidator.java [hunk 3]
```diff
@@ -53,39 +53,45 @@ public boolean validate() {
...
+                        // Regex will validate password if present
+                        if (password != null && !Pattern.compile("^"+regex+"$").matcher(password).matches()) {
+                            if(log.isDebugEnabled()) {
+                                log.debug("Regex does not match password");
+                            }
+                            this.errorType = ErrorType.INVALID_PASSWORD;
+                            return false;
+                        }
+
+                        final String username = Utils.coalesce(request.param("name"), hasParams() ? (String) param[0] : null);
```
Comment (sujithvm) [inline src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/CredentialsValidator.java:79]: Fine with keeping current behavior but username validation does not depend on regex

File: src/test/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/UserApiTest.java [hunk 4]
```diff
@@ -448,7 +449,7 @@ public void testPasswordRules() throws Exception {
         Assert.assertTrue(response.getBody().contains("xxx"));
 
         response = rh.executePutRequest("/_opendistro/_security/api/internalusers/ok1", "{\"backend_roles\":[\"my-backend-role\"],\"attributes\":{},\"password\":\"\"}", new Header[0]);
-        Assert.assertEquals(HttpStatus.SC_OK, response.getStatusCode());
+        Assert.assertEquals(HttpStatus.SC_BAD_REQUEST, response.getStatusCode());
```
Comment (sujithvm) [inline src/test/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/UserApiTest.java:452]: Do include test for empty password but fix request instead of response for the tests as they are asserting updating other attributes

File: src/test/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/UserApiTest.java [hunk 5]
```diff
@@ -534,7 +535,7 @@ public void testUserApiForNonSuperAdmin() throws Exception {
 
         // Put read only users
         response = rh.executePutRequest("/_opendistro/_security/api/internalusers/sarek", "[{ \"op\": \"add\", \"path\": \"/sarek/description\", \"value\": \"foo\" }]", new Header[0]);
-        Assert.assertEquals(HttpStatus.SC_FORBIDDEN, response.getStatusCode());
+        Assert.assertEquals(HttpStatus.SC_BAD_REQUEST, response.getStatusCode());
```
Comment (sujithvm) [inline src/test/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/UserApiTest.java:538]: Why is this bad request ? it is updating read only user and password field is not present

File: src/test/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/UserApiTest.java [hunk 6]
```diff
@@ -554,7 +555,7 @@ public void testUserApiForNonSuperAdmin() throws Exception {
 
         // Put hidden users
         response = rh.executePutRequest("/_opendistro/_security/api/internalusers/hide", "[{ \"op\": \"add\", \"path\": \"/sarek/description\", \"value\": \"foo\" }]", new Header[0]);
-        Assert.assertEquals(HttpStatus.SC_NOT_FOUND, response.getStatusCode());
+        Assert.assertEquals(HttpStatus.SC_BAD_REQUEST, response.getStatusCode());
```
Comment (sujithvm) [inline src/test/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/UserApiTest.java:558]: Why is this bad request ? it is updating reserved user and password field is not present

**Comment by sujithvm** [comment]
Please check if security kibana plugin needs to be updated to not provide empty password in any request. Admin user should be able to update attributes for other users without having to supply any password.

File: src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/CredentialsValidator.java [hunk 7]
```diff
@@ -53,39 +53,45 @@ public boolean validate() {
...
                 if (contentAsMap != null && contentAsMap.containsKey("password")) {
                     final String password = (String) contentAsMap.get("password");
 
-                    // Password can be null/empty for an existing user. Regex will validate password if present
-                    if (password != null && !password.isEmpty() && !regex.isEmpty() && !Pattern.compile("^"+regex+"$").matcher(password).matches()) {
-                        if(log.isDebugEnabled()) {
-                            log.debug("Regex does not match password");
-                        }
+                    // Password is not allowed to be empty if present.
+                    if (password != null && password.isEmpty()) {
```
Comment (sujithvm) [inline src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/CredentialsValidator.java:64]: nit: Use `isNullOrEmpty`

File: src/test/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/UserApiTest.java (hunk 5 above)
Comment (cliu123) [reply src/test/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/UserApiTest.java:538]: Moving Line #56 - 57 down [here](https://github.com/opendistro-for-elasticsearch/security/blob/master/src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/CredentialsValidator.java#L56) caused NotXContentException on Line #6. The method validate() still returns false for this use case, but bad request instead of forbidden.

File: src/test/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/UserApiTest.java (hunk 6 above)
Comment (cliu123) [reply src/test/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/UserApiTest.java:558]: Ditto

**Comment by cliu123** [comment]
> Please check if security kibana plugin needs to be updated to not provide empty password in any request. Admin user should be able to update attributes for other users without having to supply any password.

That's a good point! I'll check with Kibana team.

File: src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/CredentialsValidator.java (hunk 7 above)
Comment (cliu123) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/CredentialsValidator.java:64]: `isNullOrEmpty` is equivalent to `password != null || password.isEmpty()`, but I thought what we want to is `password != null && password.isEmpty()`.

File: src/test/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/UserApiTest.java (hunk 4 above)
Comment (cliu123) [reply src/test/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/UserApiTest.java:452]: Fixed the response for expected behavior change. And keep the original test case below.

File: src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/InternalUsersApiAction.java (hunk 2 above)
Comment (cliu123) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/InternalUsersApiAction.java]: Done.

File: src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/CredentialsValidator.java (hunk 7 above)
Comment (sujithvm) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/CredentialsValidator.java:66]: I was referring to regex null or empty but looks like that line of code got removed in this code revision

File: src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/CredentialsValidator.java [hunk 8]
```diff
@@ -62,8 +62,14 @@ public boolean validate() {
                 if (contentAsMap != null && contentAsMap.containsKey("password")) {
                     final String password = (String) contentAsMap.get("password");
 
+                    // Password is not allowed to be empty if present.
+                    if (password != null && password.isEmpty()) {
```
Comment (sujithvm) [inline src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/CredentialsValidator.java:66]: Is the empty password checked only if regex is configured ? I see lines L56-57 inserted back in.
```
&& regex != null
&& !regex.isEmpty()	           
```

Comment (cliu123) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/CredentialsValidator.java:66]: Yes, [XContentHelper.convertToMap()](https://github.com/opendistro-for-elasticsearch/security/blob/017ae5f4ec975c3324934ab58df5305297979ba7/src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/CredentialsValidator.java#L61) has dependencies on `regex != null && !regex.isEmpty()`. Without Line 56-57, it throws out [`NotXContentException`](https://github.com/opendistro-for-elasticsearch/security/blob/017ae5f4ec975c3324934ab58df5305297979ba7/src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/CredentialsValidator.java#L100) and makes the `validate()` return a false and respond bad request.

Comment (sujithvm) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/CredentialsValidator.java:66]: Could you explain the dependency of `XContentHelper.convertToMap()` on regex and why it is throwing `NotXContentException ` ? 

We would still allow users to use empty passwords unless users specify regex. Why do we want this behavior instead of rejecting empty passwords for all scenarios.

Comment (cliu123) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/CredentialsValidator.java:66]: > Could you explain the dependency of `XContentHelper.convertToMap()` on regex and why it is throwing `NotXContentException ` ?
> 
> We would still allow users to use empty passwords unless users specify regex. Why do we want this behavior instead of rejecting empty passwords for all scenarios.

There is a easy way to verify the dependencies. If you comment out the line 56 - 57 and run UserApiTest.testUserApiForNonSuperAdmin(), [this test](https://github.com/opendistro-for-elasticsearch/security/blob/4ed9f558b93082b44e80d5991d9b67c253e9f7f2/src/test/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/UserApiTest.java#L537) will get a bad response instead of a forbidden. If you check the log, you'll see the `NotXContentException`.

Comment (sujithvm) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/CredentialsValidator.java:66]: Are you saying the test code needs to be fixed? If yes, please fix as needed. 
`XContentHelper.convertToMap()` is an elasticsearch function. I don't see how it can be dependent on the regex configuration of the security plugin

Comment (cliu123) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/CredentialsValidator.java:66]: It's not dependent on the regex configuation. `XContentHelper.convertToMap()` cannot map XContent correctly when `XContent` content contains a `null` field. In this case, without `regex != null`, the regex can be null, which makes the `XContentHelper.convertToMap()` throw out [`NotXContentException`](https://github.com/elastic/elasticsearch/blob/724b244d7a72e18b9bd184f4bcb9cfc1f4967a16/server/src/main/java/org/elasticsearch/common/compress/CompressorFactory.java#L49)

Comment (sujithvm) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/CredentialsValidator.java:66]: Regex is not part of content hence it cannot interfere with `XContentHelper.convertToMap()`

Comment (sujithvm) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/CredentialsValidator.java:66]: It is converting array PATCH array body with PUT request in the test `testUserApiForNonSuperAdmin` and is failing to convert array into map. Need to fix the tests

Comment (cliu123) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/CredentialsValidator.java:64]: Ok. That's the culprit. I fixed it.

File: src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/CredentialsValidator.java (hunk 7 above)
Comment (cliu123) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/CredentialsValidator.java:65]: > nit: Use `isNullOrEmpty`

Done

File: src/test/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/UserApiTest.java (hunk 6 above)
Comment (cliu123) [reply src/test/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/UserApiTest.java]: Fixed the test.

File: src/test/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/UserApiTest.java (hunk 5 above)
Comment (cliu123) [reply src/test/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/UserApiTest.java]: Fixed the test.

File: src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/CredentialsValidator.java (hunk 3 above)
Comment (cliu123) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/CredentialsValidator.java:80]: I see.

File: src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/CredentialsValidator.java [hunk 9]
```diff
@@ -53,39 +54,45 @@ public boolean validate() {
...
+                            }
+                            return false;
                         }
-                        return false;
-                    }
 
-                    if (username.toLowerCase().equals(password.toLowerCase())) {
-                        if (log.isDebugEnabled()) {
-                            log.debug("Username must not match password");
+                        if (username.toLowerCase().equals(password.toLowerCase())) {
```
Comment (sujithvm) [inline src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/CredentialsValidator.java:89]: Can password be null here? 
Maybe check for password is not null directly in the contentAsMap as these conditions are based on non-null password entry validation

Comment (sujithvm) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/CredentialsValidator.java:89]: Also not dependent on regex but it is inside `Strings.isNullOrEmpty(regex)` block? I understand the current behavior validates password against username only if regex is defined. Just looking for improvements here. Any thoughts?

**Review by sujithvm — DISMISSED** [review]

File: src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/CredentialsValidator.java (hunk 9 above)
Comment (cliu123) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/CredentialsValidator.java:89]: Yes, I thought about that too. In theory, password has nothing to do with regex. I feel like there're some points/use cases to validate password against username no matter regex is null/empty or not and refactoring could be needed here. It's worth to discuss with the team and to get more inputs.

Comment (sujithvm) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/CredentialsValidator.java:89]: Can we address in this PR itself as change is happening around it

Comment (cliu123) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/CredentialsValidator.java:89]: I'll create another PR to make this change given that this PR has been sitting there for more than 3 weeks and it has met the goal. Let's stick with the original scope of PRs.

File: src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/CredentialsValidator.java [hunk 10]
```diff
@@ -53,39 +54,45 @@ public boolean validate() {
...
                 if (contentAsMap != null && contentAsMap.containsKey("password")) {
                     final String password = (String) contentAsMap.get("password");
 
-                    // Password can be null/empty for an existing user. Regex will validate password if present
-                    if (password != null && !password.isEmpty() && !regex.isEmpty() && !Pattern.compile("^"+regex+"$").matcher(password).matches()) {
-                        if(log.isDebugEnabled()) {
-                            log.debug("Regex does not match password");
-                        }
+                    // Password is not allowed to be empty if present.
+                    if (password != null && password.isEmpty()) {
```
Comment (vrozov) [inline src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/CredentialsValidator.java:65]: Under what condition `password` will be `null` here? It is executed with `contentAsMap.containsKey("password")` condition.

Comment (cliu123) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/CredentialsValidator.java:65]: In theory, contentAsMap.containsKey() prevents null password. But before checking isEmpty(), it's good to ensure the object "password" exists, isn't it?

Comment (vrozov) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/CredentialsValidator.java:65]: I don't see why two checks are necessary. The check `if (contentAsMap != null && contentAsMap.containsKey("password"))` looks redundant to me.

File: src/test/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/UserApiTest.java [hunk 11]
```diff
@@ -410,6 +410,7 @@ public void testPasswordRules() throws Exception {
         addUserWithPassword("ok2", "$aA123456789", HttpStatus.SC_CREATED);
         addUserWithPassword("ok3", "$Aa123456789", HttpStatus.SC_CREATED);
         addUserWithPassword("ok4", "$1aAAAAAAAAA", HttpStatus.SC_CREATED);
+        addUserWithPasswordAndHash("empty_password", "", "$%^123", HttpStatus.SC_BAD_REQUEST);
```
Comment (vrozov) [inline src/test/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/UserApiTest.java:413]: Is there a test that adds user with empty password and no hash?

File: src/test/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/UserApiTest.java [hunk 12]
```diff
@@ -448,7 +449,10 @@ public void testPasswordRules() throws Exception {
         Assert.assertTrue(response.getBody().contains("xxx"));
 
         response = rh.executePutRequest("/_opendistro/_security/api/internalusers/ok1", "{\"backend_roles\":[\"my-backend-role\"],\"attributes\":{},\"password\":\"\"}", new Header[0]);
-        Assert.assertEquals(HttpStatus.SC_OK, response.getStatusCode());
+        Assert.assertEquals(HttpStatus.SC_BAD_REQUEST, response.getStatusCode());
+
+        response = rh.executePutRequest("/_opendistro/_security/api/internalusers/non-empty_password", "{\"backend_roles\":[\"my-backend-role\"],\"attributes\":{},\"password\":\"Admin_123\"}", new Header[0]);
```
Comment (vrozov) [inline src/test/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/UserApiTest.java:454]: The request creates new user. Please change it to update existing user "ok1".

File: src/test/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/UserApiTest.java (hunk 11 above)
Comment (vrozov) [reply src/test/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/UserApiTest.java:413]: Will be also good to test request to create user with `"password": null`

File: src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/CredentialsValidator.java (hunk 10 above)
Comment (cliu123) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/CredentialsValidator.java:65]: If `contentAsMap` is `null` or `password` field doesn't exist, there's no need to validate password at all.

File: src/test/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/UserApiTest.java (hunk 11 above)
Comment (cliu123) [reply src/test/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/UserApiTest.java:413]: > Is there a test that adds user with empty password and no hash?

Line 451-452

File: src/test/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/UserApiTest.java (hunk 12 above)
Comment (cliu123) [reply src/test/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/UserApiTest.java:454]: This intends to test user creation. Line 451-452 is to test updating existing user.

File: src/test/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/UserApiTest.java (hunk 11 above)
Comment (cliu123) [reply src/test/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/UserApiTest.java:414]: Done.

File: src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/CredentialsValidator.java (hunk 10 above)
Comment (vrozov) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/CredentialsValidator.java:65]: How `contentAsMap` can be `null`? Does not check `password != null` supersede `contentAsMap.containsKey("password")`?

File: src/test/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/UserApiTest.java (hunk 12 above)
Comment (vrozov) [reply src/test/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/UserApiTest.java:456]: User creation is tested using `addUserWithPassword()`, how this test is different?

File: src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/CredentialsValidator.java (hunk 10 above)
Comment (cliu123) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/CredentialsValidator.java]: I removed `password != null`.

File: src/test/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/UserApiTest.java (hunk 12 above)
Comment (cliu123) [reply src/test/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/UserApiTest.java]: I change it to update the existing user "ok1".

File: src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/CredentialsValidator.java (hunk 10 above)
Comment (vrozov) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/CredentialsValidator.java]: Will it be better to remove `contentAsMap != null && contentAsMap.containsKey("password")`?

Comment (cliu123) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/CredentialsValidator.java]: If removing `contentAsMap != null && contentAsMap.containsKey("password")`, we'd need to keep `if (password != null && password.isEmpty()) {` at [Line 65](https://github.com/opendistro-for-elasticsearch/security/blob/1bbf8e2f6d81c5fd7f28a861ba2d56785c630405/src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/CredentialsValidator.java#L65) and add `password != null` [Line 89](https://github.com/opendistro-for-elasticsearch/security/blob/1bbf8e2f6d81c5fd7f28a861ba2d56785c630405/src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/CredentialsValidator.java#L89). Will this look better? If so, I'll update it.

Comment (vrozov) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/CredentialsValidator.java]: Replace `contentAsMap != null && contentAsMap.containsKey("password")` with 
```
String password = contentAsMap.get("password");
if (password != null) {
```

Comment (cliu123) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/CredentialsValidator.java]: Done.

**Review by vrozov — APPROVED** [review]

File: src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/CredentialsValidator.java [hunk 13]
```diff
@@ -53,39 +54,44 @@ public boolean validate() {
...
                 final Map<String, Object> contentAsMap = XContentHelper.convertToMap(this.content, false, XContentType.JSON).v2();
-                if (contentAsMap != null && contentAsMap.containsKey("password")) {
-                    final String password = (String) contentAsMap.get("password");
-
-                    // Password can be null/empty for an existing user. Regex will validate password if present
-                    if (password != null && !password.isEmpty() && !regex.isEmpty() && !Pattern.compile("^"+regex+"$").matcher(password).matches()) {
-                        if(log.isDebugEnabled()) {
-                            log.debug("Regex does not match password");
-                        }
+                String password = (String) contentAsMap.get("password");
```
Comment (sujithvm) [inline src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/CredentialsValidator.java:61]: nit: use final same as original `final String password = (String) contentAsMap.get("password");`

Comment (sujithvm) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/CredentialsValidator.java:61]: Can `contentAsMap` be null? I see the null check is removed

Comment (cliu123) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/CredentialsValidator.java:61]: `final Map<String, Object> contentAsMap = XContentHelper.convertToMap(this.content, false, XContentType.JSON).v2();` prevents null `contentAsMap`.

Comment (sujithvm) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/validation/CredentialsValidator.java:61]: Does `prevents null` mean it throws null pointer exception or returns empty map?

**Review by sujithvm — APPROVED** [review]
